fn/result.go

```diff
@@ -187,16 +187,22 @@ func AndThen[A, B any](r Result[A], f func(A) Result[B]) Result[B] {
 	return FlatMap(r, f)
 }
 
-// AndThen2 applies a function that returns a Result[C] to the success values
-// of two Result types if both exist.
-func AndThen2[A, B, C any](ra Result[A], rb Result[B],
-	f func(A, B) Result[C]) Result[C] {
-
-	return AndThen(ra, func(a A) Result[C] {
-		return AndThen(rb, func(b B) Result[C] {
-			return f(a, b)
-		})
-	})
+// LiftA2Result lifts a two-argument function to a function that can operate
+// over results of its arguments.
+func LiftA2Result[A, B, C any](f func(A, B) C,
```

Comment:
Recognize that this is quite subjective nitting, but this is quite an odd function name imo - `ToResultFn`?

If there's some canonical terminology for this that I'm missing I'll hush, but at least from a rustlang perspective I don't think there is?

Reply:
[liftA2](https://hackage.haskell.org/package/base-4.20.0.1/docs/Prelude.html#v:liftA2) is a general combinator for lifting arbitrary binary operations over any applicative functor context. The applicative here is `Result` and the intuition is that you are "lifting" some binary ("2") function `f` over the Applicative ("A"). Since go does not support HKTs, it is impossible to define this combinator once for all Applicatives, so I postfix the name "LiftA2" with the functor it is monomorphized to (in this case "Result").

There are corresponding functions for all Applicative functors, notably: Option, Result, List (two kinds: cartesian product and zip lists), Futures, functions *from* some fixed type, and any composition of these functors, since Applicatives are closed under composition (if `F` and `G` are both applicative functors then `F . G` and `G . F` are themselves applicative functors).

As with most things I'm not attached to the names, but I did choose this name for a reason. You are not the first person to provide this feedback and so I'm happy to rename this lifting combinator, but I think it's important that the name chosen is consistent for all of its instantiations (There is currently a `LiftA2Option` that would also need to be renamed).

Reply:
> As with most things I'm not attached to the names, but I did choose this name for a reason. You are not the first person to provide this feedback and so I'm happy to rename this lifting combinator, but I think it's important that the name chosen is consistent for all of its instantiations (There is currently a LiftA2Option that would also need to be renamed).

Happy to leave name as-is and keep it consistent, thanks for explaining the reasoning!

Reply:
I like LiftA2Result name